### ISSUE TITLE
Segfaulting on alpine / postgresql 9.6.0-r1

### ISSUE THREAD

**Issue by drnic** [zalando/patroni#341]
TL;DR - see "Update 4" - the issue seems to be introduced with alpine package postgresql 9.6.0-r1.

My alpine image with patroni/wal-e was running ok yesterday; and when I rebuilt it today it started segfaulting. Annoyingly, afaict, I haven't changed any top-level packages and all my Docker build logs have scrolled off the top of my iTerm screen so I cannot see what was in the image previously (and the host docker machine has been recreated since then).

I built a version of python3 with gdb support but it unhelpfully is saying that its the startup program that's failing, rather than something internally that can be inspected:

```
bash-4.3# gdb -ex r --args python3 /patroni/patroni.py
GNU gdb (GDB) 7.11
Copyright (C) 2016 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-alpine-linux-musl".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from python3...done.
Starting program: /usr/local/bin/python3 /patroni/patroni.py
warning: Error disabling address space randomization: Operation not permitted
During startup program terminated with signal SIGSEGV, Segmentation fault.
```

Currently installed alpine packages:

```
# apk info | xargs -I % apk info % | grep description: | awk '{print $1}' | sort
abuild-2.27.1-r1
abuild-2.28.0-r3
acct-6.6.2-r0
acct-6.6.2-r0
alpine-baselayout-3.0.3-r0
alpine-baselayout-3.0.3-r2
alpine-keys-1.1-r0
alpine-keys-1.2-r0
alpine-sdk-0.4-r3
alpine-sdk-0.5-r0
apk-tools-2.6.7-r0
apk-tools-2.6.7-r2
attr-2.4.47-r4
attr-2.4.47-r4
bash-4.3.42-r3
bash-4.3.46-r3
binutils-2.26-r0
binutils-2.27-r0
binutils-libs-2.26-r0
binutils-libs-2.27-r0
build-base-0.4-r1
build-base-0.4-r1
busybox-1.24.2-r11
busybox-1.25.1-r0
bzip2-1.0.6-r5
bzip2-1.0.6-r5
ca-certificates-20160104-r4
ca-certificates-20160104-r6
cdrkit-1.1.11-r2
cdrkit-1.1.11-r2
cryptsetup-libs-1.7.1-r0
cryptsetup-libs-1.7.2-r1
curl-7.50.3-r0
curl-7.50.3-r1
db-5.3.28-r0
db-5.3.28-r0
device-mapper-2.02.154-r0
device-mapper-2.02.162-r0
expat-2.1.1-r1
expat-2.2.0-r0
fakeroot-1.20.2-r3
fakeroot-1.21-r0
file-5.27-r0
file-5.28-r0
fortify-headers-0.8-r0
fortify-headers-0.8-r0
g++-5.3.0-r0
g++-6.2.1-r0
gcc-5.3.0-r0
gcc-6.2.1-r0
gdal-2.1.0-r1
gdbm-1.11-r1
gdbm-1.12-r0
geos-3.5.0-r0
giflib-5.1.4-r1
giflib-5.1.4-r1
git-2.10.1-r0
git-2.8.3-r0
gmp-6.1.0-r0
gmp-6.1.1-r0
go-1.6.3-r0
isl-0.14.1-r0
isl-0.17.1-r0
jq-1.5-r1
jq-1.5-r3
kmod-22-r0
kmod-23-r0
lddtree-1.25-r2
lddtree-1.25-r2
libatomic-5.3.0-r0
libatomic-6.2.1-r0
libattr-2.4.47-r4
libattr-2.4.47-r4
libblkid-2.28-r3
libblkid-2.28.2-r0
libbz2-1.0.6-r5
libbz2-1.0.6-r5
libc-dev-0.7-r0
libc-dev-0.7-r1
libc-utils-0.7-r0
libc-utils-0.7-r1
libcap-2.25-r0
libcap-2.25-r1
libcrypto1.0-1.0.2i-r0
libcrypto1.0-1.0.2j-r0
libcrypto1.0-1.0.2j-r1
libcurl-7.50.3-r0
libcurl-7.50.3-r1
libedit-20150325.3.1-r3
libedit-20150325.3.1-r3
libffi-3.2.1-r2
libffi-3.2.1-r2
libgcc-5.3.0-r0
libgcc-6.2.1-r0
libgomp-5.3.0-r0
libgomp-6.2.1-r0
libjpeg-turbo-1.4.2-r0
libjpeg-turbo-1.5.1-r0
libldap-2.4.44-r1
libldap-2.4.44-r3
libmagic-5.27-r0
libmagic-5.28-r0
libpng-1.6.21-r0
libpng-1.6.25-r0
libpq-9.5.4-r0
libpq-9.6.0-r1
libressl-dev-2.4.3-r1
libressl2.4-libcrypto-2.4.3-r1
libressl2.4-libssl-2.4.3-r1
libressl2.4-libtls-2.4.3-r1
libsasl-2.1.26-r7
libsasl-2.1.26-r8
libssh2-1.7.0-r0
libssh2-1.7.0-r2
libssl1.0-1.0.2i-r0
libssl1.0-1.0.2j-r0
libssl1.0-1.0.2j-r1
libstdc++-5.3.0-r0
libstdc++-6.2.1-r0
libuuid-2.28-r3
libuuid-2.28.2-r0
libxml2-2.9.4-r0
libxml2-2.9.4-r0
linux-headers-4.4.6-r1
linux-headers-4.4.6-r1
lzo-2.09-r1
lzo-2.09-r1
make-4.1-r1
make-4.2.1-r0
mkinitfs-3.0.5-r1
mkinitfs-3.0.5-r4
mpc1-1.0.3-r0
mpc1-1.0.3-r0
mpfr3-3.1.2-r0
mpfr3-3.1.5-r0
mtools-4.0.18-r1
mtools-4.0.18-r1
musl-1.1.14-r12
musl-1.1.15-r3
musl-dev-1.1.14-r12
musl-dev-1.1.15-r3
musl-utils-1.1.14-r12
musl-utils-1.1.15-r3
ncurses-libs-6.0-r7
ncurses-libs-6.0-r7
ncurses-terminfo-6.0-r7
ncurses-terminfo-6.0-r7
ncurses-terminfo-base-6.0-r7
ncurses-terminfo-base-6.0-r7
openssl-1.0.2j-r0
openssl-1.0.2j-r1
patch-2.7.5-r1
patch-2.7.5-r1
pax-utils-1.1.6-r0
pax-utils-1.1.6-r0
pcre-8.38-r1
pcre-8.39-r0
pkgconf-0.9.12-r0
pkgconf-1.0.1-r0
pkgconf-1.0.1-r0
pkgconfig-0.25-r1
postgresql-9.5.4-r0
postgresql-9.6.0-r1
postgresql-bdr-9.4.5_p1-r3
postgresql-bdr-9.4.9_p1-r0
postgresql-client-9.5.4-r0
postgresql-client-9.6.0-r1
postgresql-contrib-9.5.4-r0
postgresql-contrib-9.6.0-r1
postgresql-dev-9.5.4-r0
postgresql-dev-9.6.0-r1
postgresql-libs-9.5.4-r0
postgresql-libs-9.6.0-r1
proj4-4.9.1-r0
py-pip-8.1.2-r0
py-setuptools-20.8.0-r0
py-setuptools-28.0.0-r0
py2-pip-8.1.2-r0
python-2.7.12-r0
python2-2.7.12-r4
python3-3.5.2-r0
python3-3.5.2-r4
python3-dev-3.5.2-r0
python3-dev-3.5.2-r4
readline-6.3.008-r4
readline-6.3.008-r4
scanelf-1.1.6-r0
scanelf-1.1.6-r0
sed-4.2.2-r0
sed-4.2.2-r0
sqlite-libs-3.13.0-r0
sqlite-libs-3.14.2-r0
squashfs-tools-4.3-r2
squashfs-tools-4.3-r2
sudo-1.8.16-r0
sudo-1.8.18-r0
tar-1.29-r0
tar-1.29-r0
tiff-4.0.6-r3
tiff-4.0.6-r3
xz-libs-5.2.2-r1
xz-libs-5.2.2-r1
zlib-1.2.8-r2
zlib-1.2.8-r2
```

Installed python3/pip3 packages:

```
+ pip3 list
awscli (1.11.3)
azure (1.0.3)
azure-common (1.1.4)
azure-mgmt (0.20.2)
azure-mgmt-common (0.20.0)
azure-mgmt-compute (0.20.1)
azure-mgmt-network (0.20.1)
azure-mgmt-nspkg (1.0.0)
azure-mgmt-resource (0.20.1)
azure-mgmt-storage (0.20.0)
azure-nspkg (1.0.0)
azure-servicebus (0.20.1)
azure-servicemanagement-legacy (0.20.2)
azure-storage (0.20.3)
Babel (2.3.4)
boto (2.42.0)
botocore (1.4.61)
click (6.6)
colorama (0.3.7)
debtcollector (1.8.0)
dnspython (1.15.0)
dnspython3 (1.15.0)
docutils (0.12)
envdir (0.7)
futures (3.0.5)
gcloud (0.18.3)
gevent (1.1.2)
googleapis-common-protos (1.3.5)
greenlet (0.4.10)
httplib2 (0.9.2)
iso8601 (0.1.11)
jmespath (0.9.0)
kazoo (2.2.1)
keystoneauth1 (2.13.0)
monotonic (1.2)
msgpack-python (0.4.8)
netaddr (0.7.18)
netifaces (0.10.5)
oauth2client (3.0.0)
oslo.config (3.17.0)
oslo.i18n (3.9.0)
oslo.serialization (2.13.0)
oslo.utils (3.16.0)
pbr (1.10.0)
pgxnclient (1.2.1)
pip (8.1.2)
positional (1.1.1)
prettytable (0.7.2)
protobuf (3.1.0.post1)
psycopg2 (2.6.2)
pyasn1 (0.1.9)
pyasn1-modules (0.0.8)
pyparsing (2.1.10)
python-consul (0.6.0)
python-dateutil (2.5.3)
python-etcd (0.4.3)
python-keystoneclient (3.6.0)
python-swiftclient (3.1.0)
pytz (2016.7)
PyYAML (3.12)
requests (2.11.1)
rfc3986 (0.4.1)
rsa (3.4.2)
s3transfer (0.1.8)
setuptools (28.3.0)
six (1.10.0)
stevedore (1.17.1)
tzlocal (1.2.2)
urllib3 (1.18)
wal-e (1.0.0)
wrapt (1.10.8)
```

**Could anyone with a working alpine/patroni image, please run the following and do a diff with the above?** Thanks!

```
apk info | xargs -I % apk info % | grep description: | awk '{print $1}' | sort
pip3 list
```

I've tried opening `python3` REPL and importing each of the packages referenced by patroni source code but no errors.

Any protips on how I can debug from here?

Anyone else who built their patroni image today from scratch have a similar issue (on alpine or any other distro)?
### Update 1:

Wrote a script to look up all the pip3 packages/versions via pypi API to find the `upload_time`, and the following packages were uploaded in last day or so:

```
2016-10-12T23:23:34 awscli 1.11.3
2016-10-12T23:23:22 botocore 1.4.61
2016-10-12T23:23:43 s3transfer 0.1.8
```
### Update 2:

Restricted all three to their prior versions but still segfaulting.

```
2016-10-06T23:54:01 awscli 1.11.2
2016-10-06T23:53:49 botocore 1.4.60
2016-10-06T23:54:11 s3transfer 0.1.7
```
### Update 3:

AFAIK there is no API for https://pkgs.alpinelinux.org/packages so I paginated thru a dozen pages until the release dates were early 2016-10-11 and couldn't visually spot an alpine package that had been updated that matched one of those above. `python3.4` `3.4.3-r1` came out 2016-10-11; but I have `python3.5` installed above.

**Comment by drnic** [comment]
### Update 4: Downgraded to postgresql 9.5 - no more segfault

Even though I've been trialing PG 9.6 successfully, I just noticed that alpine pkg 9.6.0-r1 was touched recently https://pkgs.alpinelinux.org/packages?name=postgresql&branch=&repo=&arch=&maintainer=

Downgraded to PG 9.5.4 for now and no more segfaults.

**Comment by alexeyklyukin** [comment]
@drnic is this still an issue for you with the already released PostreSQL 9.6.1?

**Comment by drnic** [comment]
This issue with 9.6.0 is resolved when trying 9.6.2
